File: app/src/main/java/com/chiranths/jobportal1/Activities/BasicActivitys/ConstructionActivity.java
package com.chiranths.jobportal1.Activities.BasicActivitys;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;

import com.chiranths.jobportal1.Adapters.ConstructorAdaptor;
import com.chiranths.jobportal1.Adapters.TravelsAdaptor;
import com.chiranths.jobportal1.Model.ConstructionModel;
import com.chiranths.jobportal1.Model.TravelsModel;
import com.chiranths.jobportal1.R;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;

public class ConstructionActivity extends AppCompatActivity implements View.OnClickListener {

    CardView cv_contractors,cv_architects,cv_interior_designers,cv_construction_meterials,cv_hardware_welders,
            cv_painters,cv_carpenters,cv_electricians,cv_plumbers;
    String Name,category,cost,contactDetails,contactDetails2,experience,service1,service2,service3,service4,saveCurrentDate,saveCurrentTime,discription,vehicleNumber;
    RecyclerView rv_construction;
    ArrayList<ConstructionModel> constructioninfo =new ArrayList();
    private ConstructorAdaptor constructionAdaptor;
    Handler mHandler = new Handler();
    String type;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_construction);

        initilize();
        fetchdata();
    }

    private void initilize() {

        cv_contractors = findViewById(R.id.cv_contractors);
        cv_architects = findViewById(R.id.cv_architects);
        cv_interior_designers = findViewById(R.id.cv_interior_designers);
        cv_construction_meterials = findViewById(R.id.cv_construction_meterials);
        cv_hardware_welders = findViewById(R.id.cv_hardware_welders);
        cv_painters = findViewById(R.id.cv_carpenters);
        cv_carpenters = findViewById(R.id.cv_carpenters);
        cv_electricians = findViewById(R.id.cv_electricians);
        cv_plumbers = findViewById(R.id.cv_plumbers);

        cv_contractors.setOnClickListener(this);
        cv_architects.setOnClickListener(this);
        cv_interior_designers.setOnClickListener(this);
        cv_construction_meterials.setOnClickListener(this);
        cv_hardware_welders.setOnClickListener(this);
        cv_painters.setOnClickListener(this);
        cv_carpenters.setOnClickListener(this);
        cv_electricians.setOnClickListener(this);
        cv_plumbers.setOnClickListener(this);
        rv_construction = findViewById(R.id.rv_constructions_relates);


    }

    private void fetchdata() {
        DatabaseReference productsinfo = FirebaseDatabase.getInstance().getReference().child("constructionforyou");
        productsinfo.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()){
                    HashMap<String, Object> dataMap = (HashMap<String, Object>) dataSnapshot.getValue();
                    for (String key : dataMap.keySet()){
                        Object data = dataMap.get(key);
                        try{
                            HashMap<String, Object> userData = (HashMap<String, Object>) data;
                            constructioninfo.add(new ConstructionModel(String.valueOf(userData.get("pid")),
                                    String.valueOf(userData.get("date")),
                                    String.valueOf(userData.get("time")),
                                    String.valueOf(userData.get("name")),
                                    String.valueOf(userData.get("category")),
                                    String.valueOf(userData.get("cost")),
                                    String.valueOf(userData.get("number1")),
                                    String.valueOf(userData.get("number2")),
                                    String.valueOf(userData.get("experience")),
                                    String.valueOf(userData.get("servicess1")),
                                    String.valueOf(userData.get("servicess2")),
                                    String.valueOf(userData.get("servicess3")),
                                    String.valueOf(userData.get("servicess4")),
                                    String.valueOf(userData.get("discription")),
                                    String.valueOf(userData.get("verified")),
                                    String.valueOf(userData.get("image")),
                                    String.valueOf(userData.get("image2"))));

                        }catch (ClassCastException cce){

                            try{
                                String mString = String.valueOf(dataMap.get(key));
                                //addTextToView(mString);
                            }catch (ClassCastException cce2){

                            }
                        }
                    }

                    // Upcoming Event
                    constructionAdaptor = new ConstructorAdaptor(constructioninfo, ConstructionActivity.this);
                    RecyclerView.LayoutManager elayoutManager = new LinearLayoutManager(ConstructionActivity.this,RecyclerView.VERTICAL,false);
                    rv_construction.setLayoutManager(new GridLayoutManager(ConstructionActivity.this, 1));
                    rv_construction.setItemAnimator(new DefaultItemAnimator());
                    rv_construction.setItemAnimator(new DefaultItemAnimator());
                    mHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            rv_construction.setAdapter(constructionAdaptor);
                        }
                    });
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    @Override
    public void onClick(View view) {
        Bundle bundle = new Bundle();
        switch (view.getId()) {
            case R.id.ll_car:
                Intent intent = new Intent(this, TravelsListactivity.class);
                bundle.putString("type","car");
                intent.putExtras(bundle);
                startActivity(intent);
                break;

            case R.id.ll_bus:
                Intent intent2 = new Intent(this, TravelsListactivity.class);
                bundle.putString("type","bus");
                intent2.putExtras(bundle);
                startActivity(intent2);
                break;

            case R.id.ll_tempotravel:
                Intent intent3 = new Intent(this, TravelsListactivity.class);
                bundle.putString("type","tt");
                intent3.putExtras(bundle);
                startActivity(intent3);
                break;

            case R.id.ll_auto:
                Intent intent4 = new Intent(this, TravelsListactivity.class);
                bundle.putString("type","auto");
                intent4.putExtras(bundle);
                startActivity(intent4);
                break;

            case R.id.ll_transports:
                Intent intent5 = new Intent(this, TravelsListactivity.class);
                bundle.putString("type","transport");
                intent5.putExtras(bundle);
                startActivity(intent5);
                break;

            case R.id.ll_heavy_vehicles:
                Intent intent6 = new Intent(this, Travelsactivity.class);
                bundle.putString("type","HV");
                intent6.putExtras(bundle);
                startActivity(intent6);
                break;
        }
    }
}

File: app/src/main/java/com/chiranths/jobportal1/Activities/Construction/ConstructionActivity.kt
package com.chiranths.jobportal1.Activities.Construction

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.chiranths.jobportal1.Activities.BasicActivitys.SearchActivity
import com.chiranths.jobportal1.Adapters.BusinessCategoryAdaptor
import com.chiranths.jobportal1.Adapters.ConstructorAdaptor
import com.chiranths.jobportal1.Model.Categorymmodel
import com.chiranths.jobportal1.Model.ConstructionModel
import com.chiranths.jobportal1.R
import com.chiranths.jobportal1.Utilitys.AppConstants
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class ConstructionActivity : AppCompatActivity() {
    var backarrow : ImageView? = null
    var cv_plumbers: CardView? = null
    var Name: String? = null
    var category: String? = null
    var saveCurrentDate: String? = null
    var saveCurrentTime: String? = null
    var discription: String? = null
    var vehicleNumber: String? = null
    var llsearch: LinearLayout? = null
    var edt_const_search : EditText? = null
    var rv_construction: RecyclerView? = null
    private var rv_category: RecyclerView? = null
    var businesscatAdaptor: BusinessCategoryAdaptor? = null
    var bundle = Bundle()

    var constructioninfo: ArrayList<ConstructionModel?> = ArrayList<ConstructionModel?>()
    var categorylists = java.util.ArrayList<Categorymmodel>()

    private var constructionAdaptor: ConstructorAdaptor? = null
    var mHandler = Handler()
    var type: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_construction)
        initilize()
        fetchdata()
        fetchbusinessCategorys()
    }

    private fun initilize() {
        rv_construction = findViewById(R.id.rv_constructions_relates)
        rv_category = findViewById(R.id.id_gridview_construction)
        backarrow = findViewById(R.id.iv_const_backarrow)
        llsearch = findViewById(R.id.ll_search_business)
        edt_const_search = findViewById(R.id.edt_const_search)

        llsearch?.setOnClickListener {
            val intent = Intent(this@ConstructionActivity, SearchActivity::class.java)
            bundle.putString("searchtype", "const")
            intent.putExtras(bundle)
            startActivity(intent)
        }
        edt_const_search?.setOnClickListener {
            val intent = Intent(this@ConstructionActivity, SearchActivity::class.java)
            bundle.putString("searchtype", "const")
            intent.putExtras(bundle)
            startActivity(intent)
        }
        backarrow?.setOnClickListener {
            this.finish()
        }
    }

    private fun fetchbusinessCategorys() {
        val categorylist =
                FirebaseDatabase.getInstance().reference.child("constListing_category")
        categorylist.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    val dataMap = snapshot.value as HashMap<String, Any>?
                    for (key in dataMap!!.keys) {
                        val data = dataMap[key]
                        try {
                            val userData = data as HashMap<String, Any>?
                            categorylists.add(
                                    Categorymmodel(
                                            userData!![AppConstants.pid].toString(),
                                            userData[AppConstants.image].toString(),
                                            userData[AppConstants.category].toString(),
                                            userData["subcategory"].toString()
                                    )
                            )
                        } catch (cce: ClassCastException) {
                            try {
                                val mString = dataMap[key].toString()
                                //addTextToView(mString);
                            } catch (cce2: ClassCastException) {
                            }
                        }
                    }
                    //RecyclerView.LayoutManager nlayoutManager1 = new LinearLayoutManager(BusinessActivity.this, RecyclerView.HORIZONTAL, false);
                    val nlayoutManager1 = GridLayoutManager(this@ConstructionActivity, 4)
                    rv_category?.setLayoutManager(nlayoutManager1)
                    rv_category?.setItemAnimator(DefaultItemAnimator())
                    businesscatAdaptor =
                            BusinessCategoryAdaptor(categorylists, this@ConstructionActivity)
                    mHandler.post { rv_category?.setAdapter(businesscatAdaptor) }
                    // businesscatAdaptor.notifyItemRangeInserted(0, businesslist.size)
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun fetchdata() {
        val productsinfo = FirebaseDatabase.getInstance().reference.child("constructionforyou")
        productsinfo.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    val dataMap = dataSnapshot.value as HashMap<String, Any>?
                    for (key in dataMap!!.keys) {
                        val data = dataMap[key]
                        try {
                            val userData = data as HashMap<String, Any>?
                            if (userData?.get(AppConstants.Status).toString().equals("2")) {
                                constructioninfo.add(
                                        ConstructionModel(
                                                userData!![AppConstants.pid].toString(),
                                                userData[AppConstants.date].toString(),
                                                userData[AppConstants.time].toString(),
                                                userData["name"].toString(),
                                                userData[AppConstants.category].toString(),
                                                userData["cost"].toString(),
                                                userData["number1"].toString(),
                                                userData["product_services"].toString(),
                                                userData["experience"].toString(),
                                                userData["servicess1"].toString(),
                                                userData["servicess2"].toString(),
                                                userData["servicess3"].toString(),
                                                userData["servicess4"].toString(),
                                                userData["discription"].toString(),
                                                userData[AppConstants.verified].toString(),
                                                userData[AppConstants.image].toString(),
                                                userData[AppConstants.image2].toString(),
                                                userData["owner"].toString(),
                                                userData["address"].toString(),
                                                userData[AppConstants.Status].toString(),
                                                userData["gst"].toString(),
                                                userData["workingHrs"].toString()
                                        )
                                )
                            }
                        } catch (cce: ClassCastException) {
                            try {
                                val mString = dataMap[key].toString()
                                //addTextToView(mString);
                            } catch (cce2: ClassCastException) {
                            }
                        }
                    }

                    // Upcoming Event
                    constructionAdaptor = ConstructorAdaptor(constructioninfo, this@ConstructionActivity)
                    val elayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this@ConstructionActivity, RecyclerView.VERTICAL, false)
                    rv_construction!!.layoutManager = GridLayoutManager(this@ConstructionActivity, 1)
                    rv_construction!!.itemAnimator = DefaultItemAnimator()
                    rv_construction!!.itemAnimator = DefaultItemAnimator()
                    mHandler.post { rv_construction!!.adapter = constructionAdaptor }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }
  }